parser/src/python.lalrpop

```diff
@@ -318,6 +321,52 @@ AssertStatement: ast::Stmt = {
     },
 };
 
+LineMagicStatement: ast::Stmt = {
+    <location:@L> <m:line_magic> <end_location:@R> =>? {
+        if mode == Mode::Jupyter {
+            Ok(ast::Stmt::LineMagic(
+                ast::StmtLineMagic {
+                    kind: format!("{}", m.0),
+                    value: m.1,
+                    range: (location..end_location).into()
+                }
+            ))
+        } else {
+            Err(LexicalError {
+                error: LexicalErrorType::OtherError("line magics are only allowed in Jupyter mode".to_string()),
+                location,
+            })?
+        }
+    }
+}
+
+LineMagicExpr: ast::Expr = {
+    <location:@L> <m:line_magic> <end_location:@R> =>? {
+        if mode == Mode::Jupyter {
+            let kind = format!("{}", m.0);
+            // This should never occur as the lexer won't allow it.
+            if !matches!(kind.as_str(), "%" | "!") {
+                Err(LexicalError {
+                    error: LexicalErrorType::OtherError("expr line magics are only allowed for % and !".to_string()),
+                    location,
+                })?
+            }
```

Comment:
This is here only for sanity check as the lexer won't tokenize this as a magic command.